In [1]:
import numpy as np 
import pandas as pd

In [2]:
df = pd.concat(map(pd.read_csv, ['Brazil.csv', 'germany.csv','germany1.csv', 
                                 'Portugal1.csv', 'Spain1.csv', 'Switzerland.csv',
                                'Switzerland1.csv', 'France1.csv',
                                'Brazil1.csv', 'Netherlands.csv', 
                                'Belgium.csv', 'USA.csv', 'England.csv',
                                'Argentina.csv', 'Uruguay.csv', 
                                'Senegal.csv', 'Ecuador.csv', 'CostaRica.csv',
                                'Mexico.csv', 'France.csv', 'Denmark.csv',
                                'Australia.csv', 'Serbia.csv',
                                'Cameroon.csv', 'Ghana.csv', 'Iran.csv',
                                'Poland.csv', 'SaudI.csv', 'Tunisia.csv', 
                                 'Japan.csv', 'Croatia.csv', 'Morocco.csv',
                                'USA1.csv', 'Spain2.csv', 'Portugal2.csv', 
                                 'England2.csv', 'Belgium2.csv', 'Uruguay2.csv',
                                'Canada2.csv', 'Netherlands2.csv', 'Saudi2.csv', 
                                'Croatia2.csv', 'Cameroon2.csv', 'Denmark2.csv',
                                'Costa2.csv', 'Morocco2.csv', 'Poland2.csv',
                                'Tunisia2.csv', 'Japan2.csv', 'Argentina2.csv', 
                                 'Italy.csv', 'South_Korea.csv', 'Nigeria.csv',
                                'Colombia.csv', 'Paraguay.csv', 'Russia.csv',
                                'Algeria.csv', 'Chile.csv', 'Yugoslavia.csv',
                                'Scotland.csv', 'Romania.csv', 'Greece.csv',
                                'Honduras.csv', 'Austria.csv', 'Bulgaria.csv', 
                                 'IvoryCoast.csv', 'SouthAfrica.csv', 'RepIreland.csv',
                                'Slovenia.csv', 'Czechoslovakia.csv', 'Hungary.csv', 
                                'Egypt.csv', 'Norway.csv', 'Peru.csv',
                                'NewZealand.csv']))



In [3]:
# 158 with wc 2022 teams

In [4]:
df.shape

(251, 17)

In [5]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [6]:
df.corr()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  \
goals_z          1.000000  0.629181   0.125657      0.368589  0.207007   
xg_z             0.629181  1.000000   0.406770      0.665850  0.324571   
crosses_z        0.125657  0.406770   1.000000      0.646978  0.264161   
boxtouches_z     0.368589  0.665850   0.646978      1.000000  0.377424   
passes_z         0.207007  0.324571   0.264161      0.377424  1.000000   
progpasses_z     0.061768  0.116295   0.285345      0.245839  0.744019   
takeons_z        0.285560  0.376271   0.148366      0.252340  0.394808   
progruns_z       0.220786  0.391521   0.276753      0.396556  0.582770   
tackles_z        0.229129  0.359879   0.230044      0.399251  0.738176   
interceptions_z  0.193225  0.286579   0.233473      0.296271  0.750746   
clearances_z    -0.215297 -0.407526  -0.270660     -0.433435 -0.128805   
blocks_z         0.047036  0.056641   0.034443      0.126768  0.382457   
aerials_z        0.022483  0.016651   0.228150      0.216961 -0.022315   
fouls_z         -0.077593 -0.016097   0.078255      0.042745  0.080839   
fouled_z         0.067214  0.177733   0.176455      0.181420  0.310929   
nsxg_z           0.395148  0.652862   0.692919      0.859569  0.485552   
results         -0.605915 -0.322644  -0.016946     -0.214293 -0.150800   

                 progpasses_z  takeons_z  progruns_z  tackles_z  \
goals_z              0.061768   0.285560    0.220786   0.229129   
xg_z                 0.116295   0.376271    0.391521   0.359879   
crosses_z            0.285345   0.148366    0.276753   0.230044   
boxtouches_z         0.245839   0.252340    0.396556   0.399251   
passes_z             0.744019   0.394808    0.582770   0.738176   
progpasses_z         1.000000   0.154228    0.318069   0.499468   
takeons_z            0.154228   1.000000    0.554336   0.393583   
progruns_z           0.318069   0.554336    1.000000   0.523611   
tackles_z            0.499468   0.393583    0.523611   1.000000   
interceptions_z      0.560871   0.387894    0.426893   0.700297   
clearances_z        -0.012348  -0.125411   -0.114634  -0.225434   
blocks_z             0.343911   0.169096    0.273165   0.267605   
aerials_z            0.030006  -0.135154   -0.100141   0.010437   
fouls_z              0.220141   0.129229    0.032750   0.015906   
fouled_z             0.193522   0.345467    0.256515   0.304496   
nsxg_z               0.367309   0.397950    0.476114   0.469007   
results             -0.030736  -0.185204   -0.208865  -0.184278   

                 interceptions_z  clearances_z  blocks_z  aerials_z   fouls_z  \
goals_z                 0.193225     -0.215297  0.047036   0.022483 -0.077593   
xg_z                    0.286579     -0.407526  0.056641   0.016651 -0.016097   
crosses_z               0.233473     -0.270660  0.034443   0.228150  0.078255   
boxtouches_z            0.296271     -0.433435  0.126768   0.216961  0.042745   
passes_z                0.750746     -0.128805  0.382457  -0.022315  0.080839   
progpasses_z            0.560871     -0.012348  0.343911   0.030006  0.220141   
takeons_z               0.387894     -0.125411  0.169096  -0.135154  0.129229   
progruns_z              0.426893     -0.114634  0.273165  -0.100141  0.032750   
tackles_z               0.700297     -0.225434  0.267605   0.010437  0.015906   
interceptions_z         1.000000     -0.132053  0.312226  -0.012411  0.142906   
clearances_z           -0.132053      1.000000  0.195803   0.282838  0.209973   
blocks_z                0.312226      0.195803  1.000000   0.022957  0.184690   
aerials_z              -0.012411      0.282838  0.022957   1.000000  0.118114   
fouls_z                 0.142906      0.209973  0.184690   0.118114  1.000000   
fouled_z                0.324582      0.028137  0.081106   0.004572  0.285976   
nsxg_z                  0.412600     -0.420793  0.156530   0.147478  0.095522   
results                -0.117818      0.118327 -0.065915  -0.032400  0.095937   

                 fo

In [7]:
df.shape

(251, 17)

In [8]:
df.duplicated().value_counts()

False    251
dtype: int64

In [9]:
df.duplicated().value_counts()

False    251
dtype: int64

# convert to csv

In [10]:
df.to_csv('WC-stats.csv', index=False)


In [11]:
from sklearn.linear_model import LinearRegression
from sklearn.inspection import permutation_importance
from sklearn.model_selection import train_test_split
from sklearn.linear_model import Ridge



In [12]:
df.head()

goals_z      xg_z  crosses_z  boxtouches_z  passes_z  progpasses_z  \
0  0.423077  0.146923  -0.136154     -0.030000  0.429231      0.037692   
1  0.216923  0.348462   0.031538      0.158462  0.835385      0.626923   
2  0.113846  0.392308   0.599231      0.513846  0.833077      0.405385   
3  0.479231  0.609231   0.227692      0.450769  0.770769      0.042308   
4  0.877692  0.773846   0.428462      0.659231  0.754615      0.335385   

   takeons_z  progruns_z  tackles_z  interceptions_z  clearances_z  blocks_z  \
0   0.244615   -0.220000   0.216154         0.270000     -0.076923 -0.097692   
1   0.270000    0.266923   1.143846         0.834615     -0.059231 -0.134615   
2   0.521538    1.139231   1.109231         0.859231      0.084615  0.103077   
3   0.337692    0.927692   0.506923         1.015385      0.020769  0.381538   
4   0.023077    0.638462   0.493846         0.637692     -0.117692 -0.033846   

   aerials_z   fouls_z  fouled_z    nsxg_z  results  
0  -0.020000 -0.224615  0.100769 -0.124615        2  
1   0.196923 -0.016923  0.033846  0.146923        1  
2   0.683846  0.213846  0.353846  0.786923        9  
3   0.038462  0.039231  0.022308  0.692308        5  
4   0.572308 -0.016154 -0.096923  0.890769        5

In [13]:
y = df['results']

In [14]:
X = df.drop(columns=['results'])

In [15]:
X_train, X_test, y_train, y_test = train_test_split(
     X, y, test_size=0.20, random_state=42)

In [16]:
model = Ridge(alpha=1e-2).fit(X_train, y_train)
model.score(X, y)

0.399474142489072

In [17]:
r = permutation_importance(model, X, y, n_repeats=30, random_state=0)

# Permutation feature importance

In [18]:
scoring = ['r2', 'neg_mean_absolute_percentage_error', 'neg_mean_squared_error']
r_multi = permutation_importance(
     model, X, y, n_repeats=30, random_state=0, scoring=scoring)

for metric in r_multi:
     print(f"{metric}") 
     r = r_multi[metric]
     for i in r.importances_mean.argsort()[::-1]:
            if r.importances_mean[i] - 2 * r.importances_std[i] > 0:
                print(f"    {X.columns[i]:<8}"
                      f"{r.importances_mean[i]:.3f}"
                      f" +/- {r.importances_std[i]:.3f}")
                
                

r2
    goals_z 0.820 +/- 0.075
    xg_z    0.049 +/- 0.013
    crosses_z0.044 +/- 0.014
    passes_z0.042 +/- 0.012
    progruns_z0.034 +/- 0.015
    fouled_z0.026 +/- 0.012
neg_mean_absolute_percentage_error
    goals_z 0.609 +/- 0.079
    fouled_z0.039 +/- 0.017
    fouls_z 0.028 +/- 0.008
    tackles_z0.011 +/- 0.005
neg_mean_squared_error
    goals_z 61.458 +/- 5.656
    xg_z    3.635 +/- 0.998
    crosses_z3.314 +/- 1.059
    passes_z3.143 +/- 0.898
    progruns_z2.585 +/- 1.105
    fouled_z1.962 +/- 0.876


- df.shape = (158, 17)
- Ridge Regression = 0.42629810615686115
- r2
    - goals_z 1.282 +/- 0.149
    - progruns_z0.096 +/- 0.029
    - xg_z    0.095 +/- 0.030
    - takeons_z0.026 +/- 0.013
- neg_mean_absolute_percentage_error
    - goals_z 0.988 +/- 0.115
- neg_mean_squared_error
    - goals_z 108.981 +/- 12.672
    - progruns_z8.139 +/- 2.494
    - xg_z    8.047 +/- 2.509
    - takeons_z2.195 +/- 1.067


- df.shape = (125, 17)

- Ridge Regression = 0.4494483686544971

- r2
    - goals_z 0.982 +/- 0.123
    - progruns_z0.132 +/- 0.039
    - xg_z    0.056 +/- 0.017
    - boxtouches_z0.046 +/- 0.023
    - nsxg_z  0.039 +/- 0.013
    - aerials_z0.031 +/- 0.014
- neg_mean_absolute_percentage_error
    - goals_z 0.888 +/- 0.122
    - progruns_z0.181 +/- 0.059
- neg_mean_squared_error
    - goals_z 89.125 +/- 11.206
    - progruns_z11.964 +/- 3.505
    - xg_z    5.055 +/- 1.543
    - boxtouches_z4.180 +/- 2.062
    - nsxg_z  3.509 +/- 1.182
    - aerials_z2.782 +/- 1.244

- df.shape = (145, 17)
- Ridge Regression = 0.43985733944194594

- r2
    - goals_z 0.919 +/- 0.122
    - progruns_z0.077 +/- 0.028
    - nsxg_z  0.069 +/- 0.026
    - boxtouches_z0.034 +/- 0.015
- neg_mean_absolute_percentage_error
    - goals_z 0.817 +/- 0.115
    - progruns_z0.124 +/- 0.036
    - tackles_z0.060 +/- 0.022
    - boxtouches_z0.048 +/- 0.023
- neg_mean_squared_error
    - goals_z 78.467 +/- 10.375
    - progruns_z6.549 +/- 2.364
    - nsxg_z  5.854 +/- 2.186
    - boxtouches_z2.943 +/- 1.292

- NMAPE = the higher the better
- NMSE
- R2 =  is a statistical measure of how close the data are to the fitted regression line . It is the percentage of variation (from 0 to 1) explained by the relationship between two variables.. metric that is used to evaluate the performance of a regression-based machine learning model.